metrics-api.py

```
@@ -153,26 +144,17 @@ def metrics_page():
             headway_min = df.headway_min[df.headway_min.notnull()] # remove NaN row (first bus of the day)
             headway_min_arr.append(df.headway_min)
         except FileNotFoundError as ex:
-            return Response(json.dumps({
-                'params': params,
-                'error': f"Arrival history not found for route {route_id} on {d.isoformat()}",
-            }, indent=2), status=404, mimetype='application/json')
+            raise Exception(f"Arrival history not found for route {route_id} on {d.isoformat()}")
         except IndexError as ex:
-            return Response(json.dumps({
-                'params': params,
-                'error': f"No arrivals found for stop {start_stop_id} on route {route_id} in direction {direction_id} on {d.isoformat()}",
-            }, indent = 2), status = 404, mimetype = 'application/json')
-
+            raise Exception(f"No arrivals found for stop {start_stop_id} on route {route_id} in direction {direction_id} on {d.isoformat()}")
+    
     headway_min = pd.concat(headway_min_arr)
     waits = pd.concat(waits)
     if end_stop_id and both_stops_same_dir:
         completed_trips = pd.concat(completed_trips)
-
     if headway_min.empty:
-        return Response(json.dumps({
-            'params': params,
-            'error': f"No arrivals for stop {start_stop_id} on route {route_id}",
-        }, indent=2), status=404, mimetype='application/json')
+        # TODO: need a better error type than this
+        raise BufferError(f"No arrivals for stop {start_stop_id} on route {route_id}")
```

So we also reach this error if there are no buses running for a given timezone - I don't want to cancel the entire request in these cases (since the buses will always stop running at a certain time) so I'm raising a specific type of error. Not sure what the best one to throw was here so just raised a `BufferError` but definitely open to suggestions.

---

Can you just return an empty dataframe (or whatever data structure you use here)? I think this isn't a case when you should throw an error; it's pretty common for there to be no buses in a given time bound so it would make more sense to just say "nothing found" than give an error. Just my $0.02.

---

If this debate is too much to resolve now, you can make an issue and we can revisit it later.

---

Yeah that's a good idea - changed it to return null values